lib/Dialect/Arc/Transforms/InferStateProperties.cpp

```diff
@@ -0,0 +1,467 @@
+      : constructInput(constructInput), condition(condition), selfArg(selfArg),
+        isDisable(isDisable) {}
+
+  EnableInfo(Value input, BlockArgument condition, BlockArgument selfArg,
+             bool isDisable)
+      : EnableInfo([=](OpBuilder &) { return input; }, condition, selfArg,
+                   isDisable) {}
+
+  std::function<Value(OpBuilder &)> constructInput;
+  BlockArgument condition;
+  BlockArgument selfArg;
+  bool isDisable;
+
+  operator bool() { return constructInput && condition && selfArg; }
+};
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// Rewriter functions
+//===----------------------------------------------------------------------===//
+
+/// Take an arc and a detected reset per output value and apply it to the arc if
+/// applicable (but does not change the state ops referring to the arc).
+static LogicalResult applyResetTransformation(arc::DefineOp arcOp,
+                                              ArrayRef<ResetInfo> resetInfos) {
+  auto outputOp = cast<arc::OutputOp>(arcOp.getBodyBlock().getTerminator());
+
+  assert(outputOp.getOutputs().size() == resetInfos.size() &&
+         "required to pass the same amount of resets as outputs of the arc");
+
+  for (auto info : resetInfos) {
+    if (!info)
+      return failure();
+
+    // We can only pull out the reset to the whole arc when all the output
+    // values have the same reset applied to them.
+    // TODO: split the arcs such that there is one for each reset kind, however,
+    // that requires a cost-model to not blow up binary-size too much
+    if (!resetInfos.empty() &&
+        (info.condition != resetInfos.back().condition ||
+         info.isZeroReset != resetInfos.back().isZeroReset))
+      return failure();
+
+    // TODO: arc.state operation only supports resets to zero at the moment.
+    if (!info.isZeroReset)
+      return failure();
+  }
+
+  if (resetInfos.empty())
+    return failure();
+
+  OpBuilder builder(outputOp);
+
+  for (size_t i = 0, e = outputOp.getOutputs().size(); i < e; ++i) {
+    auto *defOp = outputOp.getOperands()[i].getDefiningOp();
+    outputOp.getOperands()[i].replaceUsesWithIf(
+        resetInfos[i].constructInput(builder),
+        [](OpOperand &op) { return isa<arc::OutputOp>(op.getOwner()); });
+
+    if (defOp && defOp->getResult(0).use_empty())
+      defOp->erase();
+  }
+
+  return success();
+}
+
+/// Transform the given state operation to match the changes done to the arc in
+/// 'applyResetTransformation' without any additional checks.
+static void setResetOperandOfStateOp(arc::StateOp stateOp,
+                                     unsigned resetConditionIndex) {
+  Value resetCond = stateOp.getInputs()[resetConditionIndex];
+  ImplicitLocOpBuilder builder(stateOp.getLoc(), stateOp);
+
+  if (stateOp.getEnable())
+    resetCond = builder.create<comb::AndOp>(stateOp.getEnable(), resetCond);
```

Comment:
Possibly me misunderstanding but this looks like it should maybe read:

```suggestion
    enableCond = builder.create<comb::AndOp>(stateOp.getEnable(), enableCond);
```

Reply:
The way the reset and enable operands work for the state op at the moment is that the reset always has precedence, i.e., in a situation where `enable=false` and `reset=true`, the register state is set to 0 (and not keep at the same value as in the previous cycle). If there is a pattern representing a 'reset' at the end of the called arc and there is already an enable operand added to the state op, then the enable should take precedence. The line you marked makes sure that the condition of the reset is adjusted such that this overlap cannot occur (and thus the enable keeps its higher precedence). The reason it is called `resetCond` is because it gets assigned the detected reset condition value from the pattern (and this is just an adjustment to it before it is added as the reset operand to the state op.

Reply:
Ah thanks, that makes sense, sorry!